### PR TITLE
sparse aggregation protocol

### PR DESCRIPTION
**Pull request by taikiy**
clippy is updated and complaining. To be rebased on top of #782.

This diff implements the sparse aggregation protocol. The input is a vector of `(contribution_value, breakdown_key)` pairs where `breakdown_key` indicates to which bucket the `contribution_value` should be added. Since the breakdown keys are secret shared, we use `check_everything` (Sierpiński triangle thing) for equality checks. A contribution value is multiplied by a share of 1 or 0 so that the value is only contributed when the breakdown_key matches the bucket index.

### CODE REVIEW

File: src/protocol/aggregation/mod.rs [hunk 1]
```diff
@@ -21,19 +24,23 @@ use crate::{
...
 pub(crate) enum Step {
     Validator,
     ConvertValueBits,
+    ConvertBreakdownKeyBits,
+    ComputeEqualityChecks,
+    CheckTimesValue,
 }
 
-/// Binary-share aggregation protocol.
+/// Binary-share aggregation protocol for a sparse breakdown key vector input.
```
Comment (richajaindce) [inline src/protocol/aggregation/mod.rs:39]: Since this is the interface of this protocol, Can you also add a comment about the input and output format i.e. an array of [conversion value, breakdown key] and also output

File: src/protocol/aggregation/mod.rs [hunk 2]
```diff
@@ -52,139 +58,229 @@ where
...
+            CV::BITS,
+            |row, i| {
+                Replicated::new(
+                    Gf2::truncate_from(row.contribution_value.left()[i]),
+                    Gf2::truncate_from(row.contribution_value.right()[i]),
+                )
+            },
+        )
+        .await?,
+        upgrade_bit_shares(
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:77]: This can be done in parallel with the line above.

File: src/protocol/aggregation/mod.rs [hunk 3]
```diff
@@ -52,139 +58,229 @@ where
...
+        .try_fold(vec![S::ZERO; num_buckets], |mut acc, bucket| async move {
+            for (i, b) in bucket.into_iter().enumerate() {
+                acc[i] += &b;
+            }
             Ok(acc)
         })
-        .await?;
-
-    Ok(aggregate)
+        .await
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:173]: I might be misunderstanding this, but this looks like it's operating across all of the streams, but I thought each stream was able to progress independently.

File: src/protocol/aggregation/mod.rs [hunk 4]
```diff
@@ -52,139 +58,229 @@ where
...
+                                                &value.to_additive_sharing_in_large_field(),
+                                                c,
+                                                record_id,
+                                            )
+                                            .await
+                                    }
+                                }),
+                        )
+                        .await
+                }
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:162]: I find this really difficult to read / digest / understand.

Is there anything we can do to improve the code readability here? Perhaps something that leads to less indentation? Maybe factor something out into another function?

File: src/protocol/aggregation/mod.rs [hunk 5]
```diff
@@ -52,139 +58,229 @@ where
...
+    let streams = seq_join(
+        ctx.active_work(),
+        stream_iter(breakdown_keys)
+            .zip(stream_iter(contribution_values))
+            .enumerate()
+            .map(|(i, (bk, v))| {
+                let eq_ctx = &equality_check_ctx;
+                let mul_ctx = &check_times_value_ctx;
+                async move {
+                    let equality_checks = check_everything(eq_ctx.clone(), i, &bk).await?;
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:138]: Wait a minute... doesn't "i" refer to the index of the record in the stream of values? Why would we check for equality against that?

**Review by benjaminsavage — CHANGES_REQUESTED** [review]
The tests pass, so the logic is clearly accurate, but I find it quite hard to follow. I would like to try to find a way to make it easier to read / digest / understand.

File: src/protocol/aggregation/mod.rs (hunk 2 above)
Comment (taikiy) [reply src/protocol/aggregation/mod.rs:77]: Changed to use `try_join` on the tuples.

File: src/protocol/aggregation/mod.rs (hunk 3 above)
Comment (taikiy) [reply src/protocol/aggregation/mod.rs:173]: I think `try_fold` applies async computations on each stream independently. cc: @martinthomson

File: src/protocol/aggregation/mod.rs (hunk 5 above)
Comment (taikiy) [reply src/protocol/aggregation/mod.rs:138]: `i` passed to `check_everything` is used as a record ID. `check_everything` uses the record ID to execute `pregenerate_all_combinations` once per record.

File: src/protocol/aggregation/mod.rs (hunk 4 above)
Comment (taikiy) [reply src/protocol/aggregation/mod.rs:162]: Extracted the multiplication part.

File: src/protocol/aggregation/mod.rs (hunk 5 above)
Comment (benjaminsavage) [reply src/protocol/aggregation/mod.rs:134]: Ah, my bad. Makes sense.

File: src/protocol/aggregation/mod.rs (hunk 3 above)
Comment (martinthomson) [reply src/protocol/aggregation/mod.rs:154]: I'm pretty sure that this will work sequentially.

File: src/protocol/aggregation/mod.rs [hunk 6]
```diff
@@ -52,139 +58,229 @@ where
...
     pub async fn aggregate() {
-        const CONTRIBUTION_BITS: u32 = 8;
-        const EXPECTED: u128 = 36;
+        type CV = Gf8Bit;
+        type BK = Gf3Bit;
 
-        const INPUT: &[u32] = &[0, 0, 18, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 10, 0, 0, 6, 0];
+        const EXPECTED: &[u128] = &[28, 0, 0, 6, 1, 0, 0, 8];
+
+        const INPUT: &[(u32, u32)] = &[
```
Comment (martinthomson) [inline src/protocol/aggregation/mod.rs:257]: I had trouble reading the ordering here.  I read this as a k-v map, where it is a v-k map instead.

File: src/protocol/aggregation/mod.rs (hunk 3 above)
Comment (benjaminsavage) [reply src/protocol/aggregation/mod.rs:154]: @akoshelev and @martinthomson - can you please advise on how to write this code for maximum streaming performance?

File: src/protocol/aggregation/mod.rs (hunk 6 above)
Comment (benjaminsavage) [reply src/protocol/aggregation/mod.rs:257]: I also got caught up on the same thing.

File: src/protocol/aggregation/mod.rs [hunk 7]
```diff
@@ -52,139 +64,241 @@ where
...
+        stream_iter(breakdown_keys)
+            .zip(stream_iter(contribution_values))
+            .enumerate()
+            .map(|(i, (bk, v))| {
+                let eq_ctx = &equality_check_ctx;
+                let c = ctx.clone();
+                async move {
+                    let equality_checks = check_everything(eq_ctx.clone(), i, &bk).await?;
+                    equality_bits_times_value(&c, equality_checks, num_buckets, v, num_records, i)
+                        .await
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:142]: My assumption is that we want to send work over to other streams right here (as opposed to in a second loop).

Right at this point (after this await) we should have N secret-shares. I think this is the place to just iterate over all of them and essentially just Send them over to other threads for processing.

I think creating a second loop *might* work, but it's confusing because it requires you to understand that the loop is being evaluate in a lazy way...

File: src/protocol/aggregation/mod.rs [hunk 8]
```diff
@@ -52,139 +64,241 @@ where
...
+        .narrow(&Step::CheckTimesValue)
+        .set_total_records(num_records);
+
+    ctx.try_join(
+        check_bits
+            .into_iter()
+            .take(num_buckets)
+            .enumerate()
+            .map(|(check_idx, check)| {
+                let step = BitOpStep::from(check_idx);
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:180]: We are kind of abusing "BitOpStep" now. This was designed for bitwise operations and cannot go above 64. There might be thousands of breakdowns - so maybe we should think more about this.

File: src/protocol/aggregation/mod.rs [hunk 9]
```diff
@@ -52,139 +64,241 @@ where
...
+                let c = check_times_value_ctx.narrow(&step);
+                let record_id = RecordId::from(record_id);
+                let v = &value_bits;
+                async move {
+                    check
+                        .multiply(&v.to_additive_sharing_in_large_field(), c, record_id)
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:186]: We should not be doing this in a loop. We should do it only once outside of the loop.

File: src/protocol/aggregation/mod.rs [hunk 10]
```diff
@@ -52,139 +64,241 @@ where
...
+    .await
 }
 
-fn get_gf2_breakdown_key_bits<CV, BK>(
+async fn upgrade_bit_shares<F, C, S, H, CV, BK>(
+    ctx: C,
     input_rows: &[SparseAggregateInputRow<CV, BK>],
-) -> Vec<BitDecomposed<Replicated<Gf2>>>
+    num_bits: u32,
+    f: H,
```
Comment (benjaminsavage) [inline src/protocol/aggregation/mod.rs:198]: Not clear from the name (f) or the comment (non-existing) what this param does

File: src/protocol/aggregation/mod.rs (hunk 8 above)
Comment (taikiy) [reply src/protocol/aggregation/mod.rs:184]: I have a PR in review that could solve this. We can create a step like BitOpStep but with a number of our choice.

File: src/protocol/aggregation/mod.rs (hunk 10 above)
Comment (taikiy) [reply src/protocol/aggregation/mod.rs:198]: Added the comment.

**Review by martinthomson — COMMENTED** [review]
OK, I got thinking...

File: src/protocol/aggregation/mod.rs (hunk 9 above)
Comment (martinthomson) [reply src/protocol/aggregation/mod.rs]: Funny.  I had the same comment, but I hadn't even thought of the performance implication, just that this function had no business dealing with the conversion.

File: src/protocol/aggregation/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply src/protocol/aggregation/mod.rs:144]: I like `binary_to_onehot()` or `bitwise_to_onehot()`

**Comment by taikiy** [comment]
Thanks for reviewing!
I updated the PR according to the comments but now I'm getting this error and don't know how to resolve...

```
higher-ranked lifetime error
could not prove `[async block@src/protocol/aggregation/mod.rs:271:63: 279:14]: std::marker::Send`
```
[code](https://github.com/private-attribution/ipa/pull/783/commits/477878fbc62c57ca9dd52e7679a6e87ae59630db#diff-6877f2b84a12df745aab6837412ad7cd69b0ceccb89b32730ffb67291e0f9737R271-R279)

I found a post on the web that solved the similar error by wrapping the function and simply returning it with `Send` bound signature, but that didn't seem to work. @martinthomson, @akoshelev, any idea how I can solve this? Thanks in advance.

**Comment by akoshelev** [comment]
> Thanks for reviewing! I updated the PR according to the comments but now I'm getting this error and don't know how to resolve...
> 
> ```
> higher-ranked lifetime error
> could not prove `[async block@src/protocol/aggregation/mod.rs:271:63: 279:14]: std::marker::Send`
> ```
> 
> [code](https://github.com/private-attribution/ipa/pull/783/commits/477878fbc62c57ca9dd52e7679a6e87ae59630db#diff-6877f2b84a12df745aab6837412ad7cd69b0ceccb89b32730ffb67291e0f9737R271-R279)
> 
> I found a post on the web that solved the similar error by wrapping the function and simply returning it with `Send` bound signature, but that didn't seem to work. @martinthomson, @akoshelev, any idea how I can solve this? Thanks in advance.

you could try https://github.com/private-attribution/ipa/blob/e3d974848542131ad298ef590335a8dd3bb903f9/src/seq_join.rs#L17 - seems to be the same issue

**Review by benjaminsavage — APPROVED** [review]
